### PR TITLE
Enable parallel::for_each for iterators returning proxy types

### PR DESCRIPTION
**Pull request by hkaiser**
- adding test
- flyby: add specializations for traits::is_value_proxy for segmented iterators

@atrantan please verify if this solves your issue

### CODE REVIEW

**Comment by atrantan** [comment]
I got this error message: 
`/home/atrantan/hpx-fork/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1065:30: error: unexpected ':' in nested name specifier; did you mean '::'?
    struct is_value_proxy<hpx:detail::local_vector_value_proxy<T, Data> >`

**Comment by atrantan** [comment]
Got this error message: 

> /home/atrantan/hpx-fork/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1065:30: error: unexpected ':' in nested name specifier; did you mean '::'?

**Comment by atrantan** [comment]
Even if I fix the error with missing `:`, it does not solve my issue.

**Comment by hkaiser** [comment]
@atrantan should be fixed now.

**Comment by hkaiser** [comment]
@atrantan Ok, please add a small reproducing use case here for me to have a closer look.

**Comment by atrantan** [comment]
Please find here a use case that corresponds to my issue : 
[https://github.com/atrantan/hpx/tree/test_value_proxy](url)

**Comment by atrantan** [comment]
Here is the commands to test the use case from Rostam:
`module load clang/trunk`
`module load cuda/8.0.44`
`module load cmake/3.6.2`
`cd path/to/hpx-build`
`cmake path/to/hpx-source -DCMAKE_CXX_COMPILER=clang++ -DHPX_WITH_CUDA=true -DHPX_WITH_CUDA_CLANG=true`
`make -j12 examples.compute.cuda.partitioned_vector`

**Comment by atrantan** [comment]
My use case compiles finally but I need this PR and PR #2425 to be merged so that I can submit a new PR.

**Comment by hkaiser** [comment]
@atrantan The first step for this PR to be merged would be for you to review and approve it.

File: hpx/parallel/algorithms/for_each.hpp [hunk 1]
```diff
@@ -45,11 +46,33 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
+            }
+
+            template <typename T>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            typename std::enable_if<
+                hpx::traits::is_value_proxy<T>::value
+            >::type
+            call(T && t)
+            {
+                auto tmp = hpx::util::invoke(proj_, std::forward<T>(t));
```
Comment (atrantan) [inline hpx/parallel/algorithms/for_each.hpp]: As @K-ballo suggested, we should replace `auto` by `typename T::value_type`

File: tests/regressions/computeapi/for_each_value_proxy.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename T>
+class test_value_proxy
+{
+public:
+
```
Comment (atrantan) [inline tests/regressions/computeapi/for_each_value_proxy.cpp]: Then add here `typedef T value_type;`

File: hpx/parallel/algorithms/for_each.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/parallel/algorithms/for_each.hpp]: I looked once more at this code. The idea behind this is that the user supplied lambda should be called by passing the value proxy itself, not the value wrapped by it. The code as it is in this PR allows for both, a user supplied lambda which takes its argument by value and by (non-const) reference. Obviously, the lambda has to be generic (or a PFO).

In conclusion, I don't think we should apply the changes suggested by @K-ballo.

Comment (atrantan) [inline hpx/parallel/algorithms/for_each.hpp]: After several compilation tests, in [https://github.com/STEllAR-GROUP/hpx/blob/master/examples/compute/cuda/partitioned_vector.cu](url) at line 32, Cuda Clang (probably in the device compilation step) seem to deduce the template argument `T` as an `int` and we want it to be a `hpx::compute::cuda::value_proxy<int>`. At the last forwarding operation (a.k.a passing the call operator of invoke_impl in [https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/util/invoke.hpp)](url) line 33), I checked that the template parameter pack `Ts...` is effectively `hpx::compute::cuda::value_proxy<int> &`. @hkaiser, any ideas?

Comment (hkaiser) [inline hpx/parallel/algorithms/for_each.hpp]: I think it's a mismatch between the types deduced on the host and on the device.

Comment (atrantan) [inline hpx/parallel/algorithms/for_each.hpp]: @hkaiser, It was that. Problem solved by modifying a little bit the following file value_proxy.hpp. Done in b961aba635572e45949869c0a5485ee79b4dd2c6

**Pull request merged**